src/jailer/src/cgroup.rs

```diff
@@ -1,25 +1,224 @@
 // Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
+use std::collections::hash_map::Entry::{Occupied, Vacant};
+use std::collections::HashMap;
 use std::ffi::OsStr;
 use std::fs::{self, File};
 use std::io::{BufRead, BufReader};
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
 use std::process;
+use std::str::FromStr;
 
 use regex::Regex;
 
 use crate::{readln_special, writeln_special, Error, Result};
 
-const PROC_MOUNTS: &str = "/proc/mounts";
+const PROC_MOUNTS: &str = if cfg!(test) {
+    "/tmp/firecracker/test/jailer/proc/mounts"
+} else {
+    "/proc/mounts"
+};
+
 const NODE_TO_CPULIST: &str = "/sys/devices/system/node/node"; // This constant should be removed once the `--node` argument is removed.
 
-pub struct Cgroup {
+// Enum identify cgroup version to be used
+#[derive(PartialEq, Copy, Clone)]
+pub enum CgroupVersion {
+    V1,
+    V2,
+}
+
+impl FromStr for CgroupVersion {
+    type Err = Error;
+
+    fn from_str(s: &str) -> std::result::Result<CgroupVersion, Self::Err> {
+        match s {
+            "1" => Ok(CgroupVersion::V1),
+            "2" => Ok(CgroupVersion::V2),
+            _ => Err(Error::CgroupInvalidVersion(s.to_string())),
+        }
+    }
+}
+
+// Holds information on a cgroup mount point discovered on the system
+struct CgroupMountPoint {
+    dir: String,
+    options: String,
+}
+
+// Allows creation of cgroups on the system for both versions
+pub struct CgroupBuilder {
+    version: CgroupVersion,
+    hierarchies: HashMap<String, PathBuf>,
+    mount_points: Vec<CgroupMountPoint>,
+}
+
+impl CgroupBuilder {
+    // Creates the builder object
+    // It will discover cgroup mount points and hierarchies configured
+    // on the system and cache the info required to create cgroups later
+    // within this hierarchies
+    pub fn new(ver: CgroupVersion) -> Result<Self> {
+        let mut b = CgroupBuilder {
+            version: ver,
+            hierarchies: HashMap::new(),
+            mount_points: Vec::new(),
+        };
+
+        // search PROC_MOUNTS for cgroup mount points
+        let f =
+            File::open(PROC_MOUNTS).map_err(|e| Error::FileOpen(PathBuf::from(PROC_MOUNTS), e))?;
+
+        // Regex courtesy of Filippo.
+        // This will match on both v1 and v2 mount points.
+        // For v1, the "ver" capture group will be empty (len = 0)
+        let re = Regex::new(
+            r"^([a-z2]*)[[:space:]](?P<dir>.*)[[:space:]]cgroup(?P<ver>2?)[[:space:]](?P<options>.*)[[:space:]]0[[:space:]]0$",
+        ).map_err(Error::RegEx)?;
+
+        let mut found = false;
```

Comment:
Since we insert into mount points or hierachies we can use that information instead of an extra `found` var.

Reply:
Indeed, will change.